### ISSUE TITLE
AMP command-handling and protocol-parsing should be on separate objects

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2810]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2810|
|Type|enhancement|
|Created|2007-08-31 18:21:32Z|
|Branch|https://github.com/twisted/twisted/tree/compose-2810|

Just a specific case of the general rule that composition is better than inheritance.

This is a necessary prerequisite for [#2529](https://github.com/twisted/twisted/issues/2529).  In order for there to be such a thing as an "extension" to the protocol, there needs to be a way to separate out different pieces of the implementation, and a way to dispatch particular commands to different objects.

Attachments:

* [patch1_streamamp_2081](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ce/3ce11065969d6f489db06446d4d074c6f9a052fc/03b98674de10d408ada303331de3cace845dc7c5) (6868 bytes) - added by hagna on 2007-09-12 18:49:33Z - AMP can do AMP(mmorpg()) and the test case tries it
* [streamamp_2081.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ce/3ce11065969d6f489db06446d4d074c6f9a052fc/16a44e300929d59637c4130ac2d18719adff1b09.patch) (6868 bytes) - added by Peaker on 2007-10-11 21:24:17Z - Reattaching the same file with an extension known by trac

<details><summary>Searchable metadata</summary>

```
trac-id__2810 2810
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_compose_2810 branches/compose-2810
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1188584492000000 1188584492000000
changetime__1199000285000000 1199000285000000
version__None None
owner__ 
cc__spiv cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|hagna set status to `assigned`|
|-|-|

So, from this:

```
class AMP(StatefulStringProtocol, Int16StringReceiver,
          _AmpParserBase):
```

to this:

```
class AMP:
```

StatefulStringProtocol and Int16StringReceiver do protocol parsing and _AmpParserBase does command dispatching, right?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This is more about what application code has to do in order to provide customized behavior relating to AMP.  The factoring of the AMP implementation may also need to change, but that is a secondary concern.

So, for example, from this:

```
class MMORPG(AMP):
    ...
```

to this:

```
class MMORPG(object):
    ...
protocol = AMP(MMORPG())
```

(although this is really just an off-the-cuff example, I'm not sure this is ultimately the correct API to satisfy this ticket).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

The existing APIs have to continue to be supported, of course.  I like JP's example, but here's a brief sketch of the APIs that I personally expect from this ticket.  These are not complete, but they should convey a general idea...

```
#!py
class IAMPMessageProcessor(Interface):
    """
    An L{IAMPMessageProcessor} provider can process messages for AMP.
    """
    def processMessage(box):
        """
        @param box: an AmpBox that contains a command, or a response to a
        previous command.
        """
    def ampConnectionMade():
        """
        ... maybe?  I don't know if these are necessary, they might be required for
        certain functionality.
        """
class MessageProcessorHelper(object):
    """
    Not the best name for this class, but this would implement the current
    default behavior of dispatching to messages; ideally with some
    nicely-documented, pluggable, internal APIs so that you could easily
    override areas of serialization or parsing without implementing everything
    yourself.  Many individual parts of these APIs are already defined within
    AMP, but not.

    Sort of like the existing _DispatchMixin, but public and supported.
    """
    implements(IAMPMessageProcessor)
class AMP(MessageProcessorHelper, ...):
    """
    The existing AMP protocol class.
    """
    def __init__(self, processor=None):
        """
        Create an AMP with the given message processor.
        """
        ...

    def processMessage(self, box):
        """
        Process a message.
        """
        if self.processor is not None:
            return self.processor.processMessage(box)
        else:
            return super(AMP, self).processMessage(box)
    def setMessageaProcessor(self, processor):
        """
        @param processor: an L{IAMPMessageProcessor} that will process messages
        sent to this transport.

        @raise ProtocolBusy: if the protocol has already been switched, then we
        are no longer handling messages and therefore have no message
        processor.
        """
        ...
```

The important thing is that these should pave the way for...
```
#!py
class MultiplexingMessageProcessor(object):
    """
    I look at a message and send it to one of several different potential
    objects.
    """
    implements(IAMPMessageProcessor)

    def __init__(self):
        self.processors = {}

    def processMessage(self, box):
        """
        Dispatch this box to an appropriate message processor based on the
        'processor' key.
        """
        return self.processors[box['processor']].processMessage(box)
```

In order to deal with protocol extensions, the AMP protocol itself would need to be upgraded to be a subclass of something similar to that.

Make sense?

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|hagna commented|
|-|-|

Yes, that makes a lot more sense.  Thanks JP and Glyph.

Would the method processMessage() from MessageProcessorHelper, whose name is not set in stone, do the same things as ampBoxReceived()?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [hagna](#note_5):
> Yes, that makes a lot more sense.  Thanks JP and Glyph.

No problem!  I hope we can give you enough help to turn you into a regular Twisted contributor :).

> Would the method processMessage() from MessageProcessorHelper, whose name is not set in stone, do the same things as ampBoxReceived()?

Yes.  In fact ```ampBoxReceived``` is a better name anyway, so let's keep that.

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|hagna commented|
|-|-|

```
class implementsMessageProcessor(amp.MessageProcessorHelper):
    pass 

```


```    
def test_helloWorldCommandUsingCompose(self):
        """
        Verify that a simple command can be sent and its response received with
        the high-level value parsing API.
        """
        c, s, p = connectedServerAndClient(
            ServerClass=amp.AMP(implementsMessageProcessor).__class__,
            ClientClass=amp.AMP(implementsMessageProcessor).__class__)
        L = []
        HELLO = 'world'
        c.sendHello(HELLO).addCallback(L.append)
        p.flush()
        self.assertEquals(L[0]['hello'], HELLO)
```

Seems to take the right direction running through the new code, but  

```
    def ampBoxReceived(self, box):
        if self.processor is not None:
            return self.processor.ampBoxReceived(box)  # new path
        else:
            return super(AMP, self).ampBoxReceived(box) # all the old tests do this
            # how many different tests should test the new path
            # before you know for sure that it works?

```

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|hagna commented|
|-|-|

So what public pluggable API should MessageProcessorHelper implement?  

The patch only includes enough to get what Glyph said working and one test case to show it.  It doesn't raise the exception when it should though.

It includes a circular reference like AMP.processor.AMP.processor.AMP.processor.  

This patch is better than nothing inasmuch as it generates feedback from you, please.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_12"></a>|@Peaker commented|
|-|-|

I gave a short look.

 * What is "class alpha"?
 * Why all the extra whitespace?
 * Is the code in MessageProcessorHelper.ampBoxReceived identical to the code below or have any changes been made?

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)|@Peaker set owner to hagna|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="hagna's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|hagna set status to `assigned`|
|-|-|

class alpha and the extra whitespace are mistakes.  

MessageProcessHelper.ampBoxReceived is the same as ampBoxReceived.

Did the test case work?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

(In [21846]) Branching to 'compose-2810'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

I am tempted to do all kinds of gold plating, adding interface declarations and a bunch more documentation, but I think this should go in for review now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Feel free to suggest improvements for other tickets - and of course if I missed something important in this one :).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set owner to @glyph|
|-|-|

OK let's go:
 * ampBoxReceived really ought to be 3 submethods. This may not be the best place to do this, though
 * I think the separation of ```CommandLocator``` and ```SimpleStringLocator``` doesn't worth it. Or maybe, ```CommandLocator``` should be a subclass of ```SimpleStringLocator```.
 * all the attributes of ```AMP``` should be declared at the top of the class (btw, the noPeerCertificate attribute is really lame).
 * _ParserHelper still offers some methods as transport, althought it's not a AMP subclass anymore. and it's still used as a transport... Oh well that's weird.
 * the expected interface of a boxSender is unclear. There's is this TODO in ```BoxDispatcher```, this is a first step. But having ```_sendBox``` as a public method isn't natural to me.
 * the fact that AMP is-a locator doesn't seem useful.
 * there are some warnings to add, but you've noticed that.

Overall, I don't really know what to say. The previous implementation was complex, and this is as complex, but in a different way. The AMP class didn't loose parent classes, but in the contrary gain 2 more parents, whereas I thought one goal was to bring that down. Maybe this is an effect of backward compatibility, I don't know.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Replying to [therve](#note_21):
_OK let's go:_
 * _ampBoxReceived really ought to be 3 submethods. This may not be the best place to do this, though_
   * It's no problem to do this, and it would make the method shorter; but each of the methods would be a private method that was only called once.
 * ''I think the separation of ```CommandLocator``` and ```SimpleStringLocator``` doesn't worth it. Or maybe, ```CommandLocator``` should be a subclass of ```SimpleStringLocator```.''
   * My goal was to very gradually move `SimpleStringLocator` to the sidelines.  I don't think it was a particularly good idea and I left it there in `AMP` mainly for compatibility.
 * ''all the attributes of ```AMP``` should be declared at the top of the class (btw, the noPeerCertificate attribute is really lame).''
   * OK
 * __ParserHelper still offers some methods as transport, althought it's not a AMP subclass anymore. and it's still used as a transport... Oh well that's weird._
   * This is a tiny helper utility.  AMP still has a dependency on some transport stuff.  The only real point to that change was to point out that, happily, that class no longer needs to inherit from AMP to do its job.
 * ''the expected interface of a boxSender is unclear. There's is this TODO in ```BoxDispatcher```, this is a first step. But having ```_sendBox``` as a public method isn't natural to me.''
   * One of the things that I was thinking about doing was making actual `Interface` objects for each of the new interfaces.  I thought this might be overkill, but maybe not; it would make it easier to describe the expected parameters everywhere.  The way I plan to respond to this comment is to add them; I'm going to put them into the `amp` module, for lack of a better place to go.  Obviously I will make `_sendBox` into `sendBox` in the process of doing so.
 * _the fact that AMP is-a locator doesn't seem useful._
   * This is entirely for compatibility.  People are going to have AMP classes that define responders, and those need to keep working.
 * _there are some warnings to add, but you've noticed that._
   * Looking through the code the only mention I've made was in `makeConnection`, but you used a plural.  Were there other warnings you'd like to suggest?

_Overall, I don't really know what to say. The previous implementation was complex, and this is as complex, but in a different way. The AMP class didn't loose parent classes, but in the contrary gain 2 more parents, whereas I thought one goal was to bring that down. Maybe this is an effect of backward compatibility, I don't know._

This 90% an effect of backwards compatibility, and maybe 10% an effect of convenience (I do not intend to deprecate the old way of doing things, merely to suggest the new one: using just AMP for everything is messy, but valid, and easy to get started with).  You will notice that AMP does not actually _use_ any of the functionality of its superclasses; its implementation is much more independent now.  Would you feel the implementation was less complex if there were a base AMP protocol with a subclass that mixed all these things together for convenience / compatibility?  (Now that I put it that way it seems kind of obvious, I will do that.)

In summary: `AMP` is-a `BoxDispatcher` because previously code would have invoked `callRemote` on it.  `AMP` is-a `CommandLocator` because previously code would have needed to define `Command.responder` methods on it.  `SimpleStringLocator` is separated into a different class becuase it's really a separate idea and shouldn't be tied together with handling of `Command` objects.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

OK, great, this is more clear to me :). My comments below:

Replying to [glyph](#note_22):
> Looking through the code the only mention I've made was in `makeConnection`, but you used a plural.  Were there other warnings you'd like to suggest?

The other one I saw was ```CommandLocator.lookupFunction.```

> Would you feel the implementation was less complex if there were a base AMP protocol with a subclass that mixed all these things together for convenience / compatibility?  (Now that I put it that way it seems kind of obvious, I will do that.)

Yep, I think that'd be more clear. Also, we'll be able to clearly define where we could make cleanups in the future.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph set owner to @therve<br>@glyph set status to `new`|
|-|-|

I think I've dealt with all the feedback.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_26"></a>|@therve set owner to @glyph|
|-|-|

First thing before I forgot:
```
these 2 objects' docstrings are not proper epytext:
    twisted.protocols.amp.IBoxSender.sendBox
    twisted.protocols.amp.BinaryBoxProtocol
```

 * the implements declarations aren't tested.
 * you've commented out the line ```self.recvd = ''``` in _switchTo. That probably means that something isn't tested, or that the previous comment is false.
 * maybe some blank lines could be inserted around (after classes docstrings for example).

I think that's pretty cool, and I have only spotted some details, so if you're confident for merging, please do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

(In [22115]) re [#2810](https://github.com/twisted/twisted/issues/2810) correct test to cover missing `recvd` line, put it back

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph commented|
|-|-|

(In [22116]) re [#2810](https://github.com/twisted/twisted/issues/2810) valid epytext

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_29"></a>|@glyph commented|
|-|-|

(In [22117]) re [#2810](https://github.com/twisted/twisted/issues/2810) more consistent use of whitespace after class docstrings

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_30"></a>|@glyph commented|
|-|-|

(In [22118]) re [#2810](https://github.com/twisted/twisted/issues/2810) just to be sure

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_31"></a>|@glyph set owner to @therve|
|-|-|

Given that I missed the uncovered line / potential regression in the last cycle, I'd like it if you gave it a final sign-off.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_32"></a>|@therve set owner to @glyph|
|-|-|

 * test_protocolSwitchLoseClientConnection has 'it shoult' in its docstring.

That's all I managed to find :). Please merge after the typo.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_33"></a>|@glyph set status to `closed`|
|-|-|

(In [22230]) Refactor AMP to separate sending and receiving binary boxes from handling them.

Author: glyph

Reviewer: therve

Fixes [#2810](https://github.com/twisted/twisted/issues/2810)

This change does not provide much in the way of new functionality, but it makes a group of previously opaque, private interfaces public and documented.  In so doing, it creates a bunch of new, cleaner tests, and opens up a number of more obvious possibilities for implementing protocol-extension features such as streaming transfers, multiplexing (multiple command-handlers on a single socket), and authentication.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
